src/github.com/matrix-org/dendrite/publicroomsapi/storage/storage.go

```diff
@@ -0,0 +1,211 @@
+	case "m.room.canonical_alias":
+		var content common.CanonicalAliasContent
+		field := &(content.Alias)
+		attrName := "canonical_alias"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.name":
+		var content common.NameContent
+		field := &(content.Name)
+		attrName := "name"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.topic":
+		var content common.TopicContent
+		field := &(content.Topic)
+		attrName := "topic"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.avatar":
+		var content common.AvatarContent
+		field := &(content.URL)
+		attrName := "avatar_url"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.history_visibility":
+		var content common.HistoryVisibilityContent
+		field := &(content.HistoryVisibility)
+		attrName := "world_readable"
+		strForTrue := "world_readable"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	case "m.room.guest_access":
+		var content common.GuestAccessContent
+		field := &(content.GuestAccess)
+		attrName := "guest_can_join"
+		strForTrue := "can_join"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	}
+
+	// If the event type didn't match, return with no error
+	return nil
+}
+
+// updateNumJoinedUsers updates the number of joined user in the database representation of the room identified by
+// a given room ID, using a given Matrix event.
+// If the event's membership is "join", increments the value, if not, decrements it.
+// Returns an error if the update failed.
+func (d *PublicRoomsServerDatabase) updateNumJoinedUsers(membershipEvent gomatrixserverlib.Event, roomID string) error {
+	membership, err := membershipEvent.Membership()
+	if err != nil {
+		return err
+	}
+
+	if membership == "join" {
+		return d.statements.incrementJoinedMembersInRoom(roomID)
+	} else if membership == "leave" || membership == "ban" {
+		return d.statements.decrementJoinedMembersInRoom(roomID)
+	} else {
+		return nil
+	}
+}
+
+// updateStringOrBooleanAttribute updates a given string or boolean attribute in
+// the database representation of the room identified by a given ID, using a given
+// string data field from content of the Matrix event triggering the update.
+// It determines the attribute's type by checking the length of the given strForTrue
+// parameter: If it's a non-empty string, it means that the attribute is a boolean
+// which will be set to true if the given data field matches it, and to false in
+// any other case. If the string is empty, it means that the attribute is a string
+// and will be updated with the field's value.
+// Returns an error if decoding the Matrix event's content or updating the attribute
+// failed.
+func (d *PublicRoomsServerDatabase) updateStringOrBooleanAttribute(
```

Comment:
This might be simpler to do as two functions, a updateStringAttribute, and an updateBooleanAttribute.